### PR TITLE
Zip latest

### LINKED ISSUE #48: combine_latest emit_on behavior
Should we buffer the `combine_latest` `emit_on` stream?

## Context
Consider the following situation, we have two streams `a` and `b`. We are going to use `combine_latest` to combine them, with an emit only on `a`. 5 entries come down from `a` then one comes down from `b` then one from `a`. Due to the lack of data from `b` none of the 5 initial `a` entries have been emitted since the `b` was missing. Now that we have `b` data do we expect it to emit 6 times (which would require the buffering of the `emit_on` stream or only once (which may violate the idea of the `emit_on` stream always emitting)?

@mrocklin @ordirules @danielballan

### PR DESCRIPTION
**Pull request by CJ-Wright**
Closes #48 
@limx0 care to look this over?

### CODE REVIEW

**Comment by CJ-Wright** [comment]
@ordirules @danielballan

File: streams/tests/test_core.py [hunk 1]
```diff
@@ -494,3 +494,39 @@ def test_partition_str():
...
+    L2 = d.sink_to_list()
+
+    a.emit(1)
+    a.emit(2)
+    b.emit('a')
+    b.emit('b')
+    a.emit(3)
+
+    assert L == [(1, 'a'), (2, 'a'), (3, 'b')]
+    assert L2 != L
```
Comment (ordirules) [inline streams/tests/test_core.py]: please be more explicit in the assert, write out what is expected, will help save mental energy in the long run
I assume it's `L2 == [(2, 'a'), (3, 'b')]`

**Comment by CJ-Wright** [comment]
Although I am more than happy to write whatever the consensus is, I'd personally rather have a whole bunch of nodes which do things that may overlap a little rather than have nodes which have many flags. This way the user only has to remember more nodes rather than having to remember more flags.
I'd like to get this in today if possible.

**Comment by ordirules** [comment]
We discussed about it off PR. Merging `combine_latest` with `zip_latest` is not a good idea, since it would force adding a buffer to `combine_latest`. I agree with the current PR and see its value.

However, we might consider discussing more complex scenarios for this stream before discussing merging. Let's say we have 3 streams:
```
Stream 1 : ----1----------2---------------------3-
Stream 2 : --------I--II-----III-------------------
Stream 3 : ---------------------------a----b------
```

What should be the output of this? Should it be:
```
(2, III, a), (2, III, b), (3 III, b)
```

or should it be recursive, like:
```
(2, I, a), (2, II, b), (2, III, a), (2, III, b), (3, III, b)
```

I would be tempted to say the former (which as far as I can tell is what the code is currently doing), since the latter could be done with (I think...):
```python
s4 = s1.zip_latest(s2)
s4.zip_latest(s3)
```

So my vote is we need to write two tests that output the two scenarios I mentioned, then I'm okay with it.
So something like:
```python
s1  = Stream()
s2 = Stream()
s3 = Stream()
s_simple = s1.zip_latest(s2, s3)
s_recursive = s3.zip_latest(s1.zip_latest(s2)) 

L_simple = s_simple.sink_to_list()

L_recursive = s_recursive.sink_to_list()

s1.emit(1)
s2.emit('I')
s2.emit("II")
s1.emit(2)
s2.emit("III")
s3.emit('a')
s3.emit('b')
s1.emit(3)
```
 Any other thoughts? Note, I haven't tested this, and output may not be correct but this is the idea.

**Comment by CJ-Wright** [comment]
I'm confused, which is the lossless stream in your example?

**Comment by ordirules** [comment]
I updated it to clarify. The output was incorrect. Also shortened. Please read above.

File: streamz/core.py [hunk 2]
```diff
@@ -330,6 +330,14 @@ def zip(self, *other):
         """ Combine two streams together into a stream of tuples """
         return zip(self, *other)
 
+    def zip_latest(self, *others):
+        """Combine multiple streams together to a stream of tuples
+
+        This will emit a new tuple of the elements from the lossless stream paired
```
Comment (ordirules) [inline streamz/core.py]: define which is lossless

File: streamz/tests/test_core.py [hunk 3]
```diff
@@ -494,3 +494,39 @@ def test_partition_str():
 def test_stream_name_str():
     source = Stream(name='this is not a stream')
     assert str(source) == '<this is not a stream; Stream>'
+
+
+def test_zip_latest():
```
Comment (ordirules) [inline streamz/tests/test_core.py]: add test for 3 streams

**Comment by ordirules** [comment]
After discussion with @CJ-Wright, to clarify, this routine makes an assumption. Only one stream is considered to be the lossless stream, and thus buffered. In `s1.zip_latest(s2, s3)`, the lossless stream is always `s1`. This is like `combine_latest` with `emit_on=[s1]` and an additional lossless buffer for `s1`. Once all streams have emitted at least once, the behaviour is exactly the same as `s1.combine_latest(s2, s3, emit_on=[s1])`.

The case of multiple lossless buffers is not considered, but @CJ-Wright will post an issue. It will require recursion through the buffers.

**Comment by CJ-Wright** [comment]
I'm going to merge this, everything passes and such. If we find the name super distasteful we can always change it again.

**Pull request merged**